src/main/java/edu/kit/compiler/semantic/DetailedNameTypeAstVisitor.java

```diff
@@ -125,210 +127,252 @@ public DataType visit(ClassNode classNode) {
+        conditionType_.ifPresent(conditionType -> {
+            if (conditionType.getType() != DataTypeClass.Boolean) {
+                semanticError(whileStatementNode, "while statement condition must be boolean");
+            }
+        });
 
         whileStatementNode.getStatement().accept(this);
 
-        return VoidType;
+        return Optional.empty();
     }
 
-    public DataType visit(ReturnStatementNode returnStatementNode) {
+    public Optional<DataType> visit(ReturnStatementNode returnStatementNode) {
         if (returnStatementNode.getResult().isPresent()) {
-            DataType resultType = returnStatementNode.getResult().get().accept(this);
-
-            if (!resultType.isCompatibleTo(expectedReturnType)) {
-                semanticError(returnStatementNode,
-                    "expression type %s does not match required return type %s",
-                    resultType.getRepresentation(stringTable),
-                    expectedReturnType.getRepresentation(stringTable)
-                );
-            }
-        } else if (expectedReturnType.getType() != DataTypeClass.Void) {
-            semanticError(returnStatementNode, "method requires a return value of type %s", expectedReturnType.getRepresentation(stringTable));
+            Optional<DataType> resultType_ = returnStatementNode.getResult().get().accept(this);
+            resultType_.ifPresent(resultType -> {
+                if (expectedReturnType.isPresent() && !resultType.isCompatibleTo(expectedReturnType.get())) {
+                    semanticError(returnStatementNode,
+                        "expression type %s does not match required return type %s",
+                        resultType.getRepresentation(stringTable),
+                        expectedReturnType.get().getRepresentation(stringTable)
+                    );
+                }
+            });
+        } else if (expectedReturnType.isPresent() && expectedReturnType.get().getType() != DataTypeClass.Void) {
+            semanticError(returnStatementNode, "method requires a return value of type %s", expectedReturnType.get().getRepresentation(stringTable));
         }
 
-        return VoidType;
+        return Optional.empty();
     }
 
-    public DataType visit(ExpressionStatementNode expressionStatementNode) {
+    public Optional<DataType> visit(ExpressionStatementNode expressionStatementNode) {
         expressionStatementNode.getExpression().accept(this);
 
-        return VoidType;
+        return Optional.empty();
     }
 
-    public DataType visit(BinaryExpressionNode binaryExpressionNode) {
-        DataType leftSideType = binaryExpressionNode.getLeftSide().accept(this);
-        DataType rightSideType = binaryExpressionNode.getRightSide().accept(this);
+    private Optional<DataType> applyResultType(ExpressionNode expressionNode, Optional<DataType> resultType_) {
+        resultType_.ifPresentOrElse(resultType -> {
+            expressionNode.setResultType(resultType);
```

Comment:
Some of your lambdas could be simplified using method references (in this case `expressionNode::setResultType`).

Reply:
I'm aware and I, personally, like them better, I just made the experience that some developers find them hard to read.